Helpers.gs

```diff
@@ -831,28 +831,43 @@ function sendSummary() {
  * @param {Number} maxRetries - How many times the function should try if it fails
  * @return {?Calendar.Event} The Calendar.Event that was added in the calendar, null if func did not complete successfully
  */
+var backoffRecoverableErrors = [
+  "Exception: Service invoked too many times in a short time",
+  "Exception: Rate Limit Exceeded",
+  "Exception: Quota Error: User Rate Limit Exceeded",
+  "Service error: Spreadsheets",
+  "Exception: User rate limit exceeded",
+  "Exception: Internal error. Please try again.",
+  "Exception: Cannot execute AddColumn because another task",
+  "Service invoked too many times in a short time:",
+  "Exception: Internal error.",
+  "Exception: Limit Exceeded: DriveApp."];
 function callWithBackoff(func, maxRetries) {
   var tries = 0;
-  var timeout = 100;
   var result;
-  while(tries < maxRetries){
+  while ( tries < maxRetries ) {
     tries++;
     try{
       result = func();
       return result;
     }
-    catch(e){
-      if (tries < maxRetries){
-        Logger.log(`Error, retrying in ${timeout}ms... [${e}]`);
-        Utilities.sleep(timeout);
-        timeout*=2; // Exponentially increase timeout
-      }
-      else {
-        Logger.log(`Error, giving up after trying ${maxRetries} times [${e}]`);
+    catch(err){
+      if ( err.message.includes("is not a function")  || !backoffRecoverableErrors.some(function(e){
+              return  err.message.toString().slice(0,e.length) == e  ;
```

Comment:
So I have to admit that I don't fully understand what this is doing (js isn't my strongest language), but I think that the essence of the `some.....slice.....` section is "if any of `backoffRecoverableErrors` is found in `err.message` (starting with the beginning - hence `slice(0`)". Is that how you would interpret it too?

A couple things I've noticed. First, here you pass in `err.message` to the `backoffRecoverableErrors.some` callback function and the example site I shared just uses `err`. Not sure if that makes a difference, but it might be what causes the "Exception: " prefix. If that's the case, then this `some` function would never match.

To solve this issue (and possibly condense our list a bit) can we instead trim it to something like:

```javascript
var backoffRecoverableErrors = [
  "Service invoked too many times in a short time",
  "Rate Limit Exceeded",
  "Internal error",
  "Service invoked too many times in a short time:"];
...
            if ( err.message.includes("is not a function")  || !backoffRecoverableErrors.some(function(e){
              return new RegExp(e, 'i').test(err.toString())
            }) ) {
...
```

This would limit our array a lot (I also removed ones that don't really apply to our script - like Drive, Spreadsheets, etc). Thoughts?

Reply:
Yes exactly. It is an equivalent of an startWith function looping over all the error messages. 

You are correct about the never matching. I misinterpreted the name of the variable.

Good idea about the condensing.

Reply:
I'm worried about the speed costs of my solution, though. Constructing a regex object for EVERY exception message in EVERY backoff exception for EVERY api call seems quite costly. And doing a regex match is probably also slow when compared to `includes`.

Maybe if we make the whole array lower/upper case and use a callback like `e.includes(err.message.toString().toLowerCase()`?

Thoughts? @jonas0b1011001 ?

Reply:
I see your point. It does make sense. Will wait for inputs from JS experts to commit changes.

Reply:
Benchmark tests: turns out `indexOf` is crazy fast, but `includes` is still ~8 times faster than regex

Reply:
So let's do

```javascript
var backoffRecoverableErrors = [
  "service invoked too many times in a short time",
  "rate limit exceeded",
  "internal error"];
...
            if ( err.message.includes("is not a function")  || !backoffRecoverableErrors.some(function(e){
              return err.message.toLowerCase().indexOf(e) > -1;
            }) ) {
...
```

Reply:
Perfect. Change committed

Reply:
@rufuswilson one final tweak - the strings in the `backoffRecoverableErrors` array will have to be lowercase as well

Reply:
Indeed